python/lsst/daf/persistence/genericAssembler.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python
+
+#
+# LSST Data Management System
+# Copyright 2016 LSST Corporation.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.    See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+import inspect
+
+def genericAssembler(dataId, componentInfo, cls):
+    """A generic assembler for butler composite datasets, that can be used when the component names match the
+    argument names in the __init__ signature, or the setter name for component objects is specified or can be
+    inferred by component name.
+
+    When determining setter names: If the setter name is specified by the policy then the genericAssembler
+    will use that to set the component into the python object. If the policy does not specify setter names the
+    genericAssembler will see if the __init__ func input argument names match the policy argument names. If
+    that does not work, and the python object has setter names that match the component name of all the object
+    then the setter name can be inferred; it will first try 'set' + <componentName>, and if that does not
+    exist it will try 'set' + <componentName>.capitalize (e.g. for component name 'foo', it will try setfoo
+    and then setFoo.) If no setter can be found for a component object, it will raise a runtime error.
+    """
+    canUseClassCtor = True
+    try:
+        args = inspect.getargspec(cls.__init__)
+    except TypeError:
+        # If the class constructor is implemented directly in C++ this exception happens, and then we can't
+        # use introspection to match input arguments.
+        canUseClassCtor = False
+    # If any componentInfo specifies a setter than do not use the constructor.
+    useSetters = any(info.setter for info in componentInfo.values())
+    if not useSetters and canUseClassCtor and set(args.args) - set(componentInfo.keys()) == set(['self']):
+        initArgs = {k: v.obj for k, v in componentInfo.iteritems()}
+        obj = cls(**initArgs)
+    else:
+        obj = cls()
+        for componentName, componentInfo in componentInfo.iteritems():
+            if componentInfo.setter is not None:
+                setter = getattr(obj, componentInfo.setter)
+            elif hasattr(obj, 'set' + componentName):
```

Comment:
Given our (evolving) coding standards, it's probably better to use `'set_' + componentName`.